### PR TITLE
[PHP] add AbstractPhpCodegen to be as the base class of PHP-related  generators

### PR DESCRIPTION
**Pull request by abcsun**

### CODE REVIEW

**Comment by wing328** [comment]
For https://github.com/swagger-api/swagger-codegen/issues/3227

**Comment by wing328** [comment]
cc @arnested

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractPhpCodegen.java [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+            additionalProperties.put(CodegenConstants.ARTIFACT_VERSION, artifactVersion);
+        }
+
+        if (additionalProperties.containsKey(VARIABLE_NAMING_CONVENTION)) {
+            this.setParameterNamingConvention((String) additionalProperties.get(VARIABLE_NAMING_CONVENTION));
+        }
+
+        // additionalProperties.put("escapedInvokerPackage", invokerPackage.replace("\\", "\\\\"));
+
+        // // make api and model doc path available in mustache template
```
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractPhpCodegen.java]: I think `apiDocPath` and `modelDocPath` are still used in PHP API client documentation (markdown) so we should keep it.

**Comment by wing328** [comment]
@abcsun when you've time, can you help do a rebase on the latest master to resolve the merge conflicts?

**Comment by abcsun** [comment]
@wing328 please review

**Comment by wing328** [comment]
I did some tests with PHP API client, Slim, Silex, Lumen and found 1 issue.

If I run the following command:
```
   java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate \
   -i modules/swagger-codegen/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml \
   -l lumen \
   -o /var/tmp/lumen/ 
```
it shows the following error:
```
[main] WARN io.swagger.codegen.languages.AbstractPhpCodegen - Return (reserved word) cannot be used as model name. Renamed to ModelReturn
[main] WARN io.swagger.codegen.languages.AbstractPhpCodegen - Return (reserved word) cannot be used as model name. Renamed to ModelReturn
[main] ERROR io.swagger.codegen.AbstractGenerator - null/api.mustache (No such file or directory)
[main] ERROR io.swagger.codegen.AbstractGenerator - can't load template null/api.mustache
Exception in thread "main" java.lang.RuntimeException: Could not generate api file for 'Fake'
	at io.swagger.codegen.DefaultGenerator.generate(DefaultGenerator.java:478)
	at io.swagger.codegen.cmd.Generate.run(Generate.java:223)
	at io.swagger.codegen.SwaggerCodegen.main(SwaggerCodegen.java:36)
Caused by: java.lang.RuntimeException: can't load template null/api.mustache
	at io.swagger.codegen.AbstractGenerator.readTemplate(AbstractGenerator.java:50)
	at io.swagger.codegen.DefaultGenerator.processTemplateToFile(DefaultGenerator.java:643)
	at io.swagger.codegen.DefaultGenerator.generate(DefaultGenerator.java:437)
	... 2 more
```

`./bin/lumen-petstore-server.sh` works fine as it's using `-t `modules/swagger-codegen/src/main/resources/lumen`.

@abcsun please let me know if you need help fixing the issue.

**Comment by abcsun** [comment]
@wing328 Forgot to set the templateDir in LumenServerCodegen, it has be fixed.

**Comment by wing328** [comment]
@abcsun thanks for the fix. It works for me now.

**Pull request merged**